### PR TITLE
Reengineer Search

### PR DESCRIPTION
**Pull request by elsantner**
- Add unit tests for search functionality
- Extract rule parser from Search
- Extract rule inversion via template method

### CODE REVIEW

File: src/main/java/menagerie/model/search/SearchRuleParser.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    String type = arg.substring(arg.indexOf(':') + 1);
+    if (type.equalsIgnoreCase("group")) {
+      rules.add(new TypeRule(TypeRule.Type.GROUP, inverted));
+    } else if (type.equalsIgnoreCase("media")) {
+      rules.add(new TypeRule(TypeRule.Type.MEDIA, inverted));
+    } else if (type.equalsIgnoreCase("image")) {
+      rules.add(new TypeRule(TypeRule.Type.IMAGE, inverted));
+    } else if (type.equalsIgnoreCase("video")) {
+      rules.add(new TypeRule(TypeRule.Type.VIDEO, inverted));
+    }
```
Comment (fragjulian) [inline src/main/java/menagerie/model/search/SearchRuleParser.java:112]: [optional] add additional else with warning log

File: src/test/java/menagerie/model/search/SearchRuleParserTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class SearchRuleParserTest {
+
+  @Test
+  void testParseIDRule() {
```
Comment (fragjulian) [inline src/test/java/menagerie/model/search/SearchRuleParserTest.java]: you could split up this test case into three separate ones

File: src/test/java/menagerie/model/search/SearchRuleParserTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    // Test parsing ID rule with greater than operator
+    rules = SearchRuleParser.parseRules("id:>123");
+    assertEquals(1, rules.size());
+    assertTrue(rules.get(0) instanceof IDRule);
+    assertEquals(123, ((IDRule) rules.get(0)).getId());
+    assertEquals(IDRule.Type.GREATER_THAN, ((IDRule) rules.get(0)).getType());
+  }
+
+  @Test
+  void testParseDateRule() {
```
Comment (fragjulian) [inline src/test/java/menagerie/model/search/SearchRuleParserTest.java]: same here

File: src/test/java/menagerie/model/search/SearchTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    items = new ArrayList<>();
+    items.add(createMediaItemMock(1, false, false));
+    items.add(createMediaItemMock(2, true, false));
+    items.add(createMediaItemMock(3, false, false));
+    items.add(createMediaItemMock(4, true, true));
+    items.add(createMediaItemMock(5, false, true));
+  }
+
+  @Test
+  void testApplyRules() {
```
Comment (fragjulian) [inline src/test/java/menagerie/model/search/SearchTest.java:40]: same here

File: src/test/java/menagerie/model/search/rules/TestItemBuilder.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package menagerie.model.search.rules;
+
+import menagerie.model.menagerie.Tag;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TestItemBuilder {
```
Comment (fragjulian) [inline src/test/java/menagerie/model/search/rules/TestItemBuilder.java:8]: neat!

File: src/test/java/menagerie/model/search/rules/TitleRuleTest.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class TitleRuleTest {
+
+  private GroupItem createGroupItem(String title) {
+    return new GroupItem(null, 1, 1, title);
+  }
+
+  @Test
+  void testAccept() {
```
Comment (fragjulian) [inline src/test/java/menagerie/model/search/rules/TitleRuleTest.java]: [optional] move the items to class fields and split up the test into 4 separate cases

**Review by fragjulian — APPROVED** [review]
LGTM! nice refactorings and test coverage

File: src/test/java/menagerie/model/search/SearchRuleParserTest.java (hunk 2 above)
Comment (elsantner) [reply src/test/java/menagerie/model/search/SearchRuleParserTest.java]: Implemented via parameterized tests now

File: src/test/java/menagerie/model/search/rules/TitleRuleTest.java (hunk 6 above)
Comment (elsantner) [reply src/test/java/menagerie/model/search/rules/TitleRuleTest.java]: Done

**Review by fragjulian — APPROVED** [review]

**Pull request merged**